proof/refine/ARM/Reply_R.thy

```diff
@@ -634,4 +618,486 @@ lemma replyRemoveTCB_invs':
+  by (clarsimp simp: obj_at'_def)
+
+lemma updateReply_corres_replyPrev:
+  "corres dc (reply_at rp and (\<lambda>s. rp \<notin> fst ` replies_with_sc s)) (reply_at' rp)
+            (set_reply_obj_ref (\<lambda>_. id) rp r1)
+            (updateReply rp (replyPrev_update (\<lambda>_. new)))"
+  apply (simp add: update_sk_obj_ref_def updateReply_def)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split[OF _ get_reply_corres])
+      apply (rule setReply_corres)
+      apply (simp add: reply_relation_def)
+  by wpsimp+
+
+lemma replyPrev_same_replyNext[simp]:
+  "replyPrev_same (replyNext_update (\<lambda>_. Some (Head scptr)) ko) ko"
+  by (clarsimp simp: replyPrev_same_def)
+
+crunches update_sk_obj_ref
+  for reply_at[wp]: "reply_at rp"
+  (wp: crunch_wps)
+
+crunches updateReply
+  for sc_obj_at'[wp]: "\<lambda>s. Q (obj_at' (P :: sched_context \<Rightarrow> bool) scp s)"
+  and reply_at'[wp]: "reply_at' rptr"
+
+lemma updateReply_replyNexts_replyPrevs:
+  "\<lbrace>\<lambda>s. \<forall>ko. ko_at' ko rptr s \<longrightarrow>
+       P (\<lambda>a. if a = rptr then replyNext_of (f ko) else replyNexts_of s a)
+         (\<lambda>a. if a = rptr then replyPrev (f ko) else replyPrevs_of s a)\<rbrace>
+   updateReply rptr f
+   \<lbrace>\<lambda>_ s. P (replyNexts_of s) (replyPrevs_of s)\<rbrace>"
+  unfolding updateReply_def
+  by wpsimp
+
+
+lemma replyNext_update_corres_Head:
+  "corres dc (reply_at rptr) (reply_at' rptr)
+   (set_reply_obj_ref reply_sc_update rptr (Some scptr))
+   (updateReply rptr (\<lambda>reply. replyNext_update (\<lambda>_. Some (Head scptr)) reply))"
+  unfolding update_sk_obj_ref_def updateReply_def
+  apply (rule corres_guard_imp)
+    apply (rule corres_split [OF set_reply_corres get_reply_corres])
+      apply (clarsimp simp: reply_relation_def)
+     apply wpsimp+
+  apply (clarsimp simp: obj_at'_def replyPrev_same_def)
+  done
+
+lemma replyNext_update_corres_Next:
+  "corres dc (reply_at rptr) (reply_at' rptr)
+   (set_reply_obj_ref reply_sc_update rptr None)
+   (updateReply rptr (\<lambda>reply. replyNext_update (\<lambda>_. Some (Next scptr)) reply))"
+    unfolding update_sk_obj_ref_def updateReply_def
+  apply (clarsimp simp: )
+  apply (rule corres_guard_imp)
+  apply (rule corres_split [OF set_reply_corres get_reply_corres])
+  apply (clarsimp simp: reply_relation_def)
+  apply wpsimp+
+  apply (clarsimp simp: obj_at'_def replyPrev_same_def)
+  done
+
+(* FIXME RT: move to Lemma bucket *)
+lemma case_list_when:
+  "(case l of
+      [] \<Rightarrow> return ()
+      | r # x \<Rightarrow> f r x)
+   = (when (l \<noteq> []) $ f (hd l) (tl l))"
+  by (clarsimp simp: list_case_If2)
+
+lemma update_sc_reply_stack_update_ko_at'_corres:
+  "corres dc
+     (sc_at ptr)
+     (ko_at' sc' ptr and (\<lambda>s. heap_ls (replyPrevs_of s) replies reply_ptr))
+     (update_sched_context ptr (sc_replies_update (\<lambda>_. reply_ptr)))
+     (setSchedContext ptr (scReply_update (\<lambda>_. replies) sc'))"
```

Comment:
I don't quite understand the type of `reply_ptr` and `replies` here. Probably they are meant to be swapped?
Also, having `(\<lambda>s. heap_ls (replyPrevs_of s) replies reply_ptr)` in the concrete assumption is probably a bit odd.  It should be an assumption to the corres rule. This is in a sense part of the state_relation, involving both the abstract and the concrete sides. What it says is that the new head and the new sc_replies queue shoulld satisfy `sc_replies_relation`.

Reply:
Yes, I've accidentally swapped the names. I'll fix that :+1:

Reply:
I believe that if we move that predicate to an assumption on the whole rule, it will generate a goal which states that this particular update satisfies `heap_ls` _for all heaps_. The current formulation will instead generate a precondition which only requires that `heap_ls` holds on `replyPrevs_of s`. This makes the assumptions weaker, and thus the lemma stronger. If I move the predicate into the assumption, then this lemma will work in cases where the update is quite harmless, but it will fail to work when proving `heap_ls` really requires knowing something about this particular `replyPrevs_of s`.

Reply:
I'll try to find the case where I use it which should hopeful clarify why it is needed.

Reply:
well, you can have `replyPrevs_of s` as the heap with `\forall s`, so it doesn't need to be for *all_heaps*, right?

Reply:
but yes, if the way you proved things does require it to be there, I suppose it won't harm to leave it in that way.

Reply:
The place where I use this is for binding sched contexts and reply objects in `bindScReply` and `bind_sc_reply`. In that case, the particular update is `setSchedContext scPtr (sc { scReply = Just replyPtr })` which is immediately preceded by `updateReply replyPtr (\reply -> reply { replyPrev = scReplyOpt })`. 

Now the state relation states that `∀p replies. sc_replies_of s p = Some replies ⟶ heap_ls (replyPrevs_of s') (scReplies_of s' p) replies` and this relates the concrete-side linked list with abstract-side lists. So fundamentally, updating the `scReply` field of some SchedContext is not safe, unless replyPtr is the head of an appropriate linked list. But then the assumption `\forall s'. heap_ls (replyPrevs_of s')  replyPtr replies` amounts to saying that `replyPtr` is the head of a linked list in all possible universes, which is not true. But it is true in the context of `bindScReply` because the preceding line does the appropriate update.

Does that clarify things at all?

Reply:
Ok, I think I see it, this is not to be considered as directly constituting `sc_replies_relation`. It is just the matter of the particular `replyPrevs_of` heap of the particular instance on the concrete side.

Well, you get this concrete side precondition from the fact that what `replyPtr` is newly linked to previously satisfied `sc_replies_relation`. But this is all in the context, outside of this lemma, I guess.

Reply:
That's it :+1: